File: solutions/typescript/linked-list/1/linked-list.ts
export class Node<T> {
  data: T;
  next: Node<T> | null;
  prev: Node<T> | null;

  constructor(data: T) {
    this.data = data;
    this.next = null;
    this.prev = null;
  }
}

export class LinkedList<T> {
  private head: Node<T> | null;
  private tail: Node<T> | null;
  private length: number;

  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  public push(data: T): void {
    const newNode = new Node(data);
    
    if (this.tail === null) {
      // Empty list
      this.head = newNode;
      this.tail = newNode;
    } else {
      newNode.prev = this.tail;
      this.tail.next = newNode;
      this.tail = newNode;
    }
    
    this.length++;
  }

  public pop(): T {
    if (this.tail === null) {
      throw new Error('Cannot pop from empty list');
    }
    
    const removedNode = this.tail;
    const data = removedNode.data;
    
    if (this.tail === this.head) {
      // Only one node in the list
      this.head = null;
      this.tail = null;
    } else {
      this.tail = removedNode.prev;
      this.tail!.next = null;
    }
    
    this.length--;
    return data;
  }

  public shift(): T {
    if (this.head === null) {
      throw new Error('Cannot shift from empty list');
    }
    
    const removedNode = this.head;
    const data = removedNode.data;
    
    if (this.head === this.tail) {
      // Only one node in the list
      this.head = null;
      this.tail = null;
    } else {
      this.head = removedNode.next;
      this.head!.prev = null;
    }
    
    this.length--;
    return data;
  }

  public unshift(data: T): void {
    const newNode = new Node(data);
    
    if (this.head === null) {
      // Empty list
      this.head = newNode;
      this.tail = newNode;
    } else {
      newNode.next = this.head;
      this.head.prev = newNode;
      this.head = newNode;
    }
    
    this.length++;
  }

  public delete(data: T): void {
    let current = this.head;
    
    while (current !== null) {
      if (current.data === data) {
        // Found the node to delete
        if (current === this.head && current === this.tail) {
          // Only node in the list
          this.head = null;
          this.tail = null;
        } else if (current === this.head) {
          // Node is the head
          this.head = current.next;
          this.head!.prev = null;
        } else if (current === this.tail) {
          // Node is the tail
          this.tail = current.prev;
          this.tail!.next = null;
        } else {
          // Node is in the middle
          current.prev!.next = current.next;
          current.next!.prev = current.prev;
        }
        
        this.length--;
        return;
      }
      
      current = current.next;
    }
  }

  public count(): number {
    return this.length;
  }
}

File: solutions/ruby/linked-list/1/linked_list.rb
class Deque
  class Node
    attr_accessor :value, :next, :prev

    def initialize(value)
      @value = value
      @next = nil
      @prev = nil
    end
  end

  def initialize
    @head = nil
    @tail = nil
    @size = 0 # Inicializamos el contador en cero
  end

  # Devuelve el número de elementos en la lista
  def count
    @size
  end

  def push(value)
    new_node = Node.new(value)
    if @tail.nil?
      @head = @tail = new_node
    else
      new_node.prev = @tail
      @tail.next = new_node
      @tail = new_node
    end
    @size += 1 # Sumamos una estación
  end

  def pop
    return nil if @tail.nil?

    removed_node = @tail
    @tail = @tail.prev

    if @tail.nil?
      @head = nil
    else
      @tail.next = nil
    end

    @size -= 1 # Restamos una estación
    removed_node.value
  end

  def unshift(value)
    new_node = Node.new(value)
    if @head.nil?
      @head = @tail = new_node
    else
      new_node.next = @head
      @head.prev = new_node
      @head = new_node
    end
    @size += 1 # Sumamos una estación
  end

  def shift
    return nil if @head.nil?

    removed_node = @head
    @head = @head.next

    if @head.nil?
      @tail = nil
    else
      @head.prev = nil
    end

    @size -= 1 # Restamos una estación
    removed_node.value
  end

  def delete(value)
    current = @head

    while current
      if current.value == value
        current.prev.next = current.next if current.prev
        current.next.prev = current.prev if current.next

        @head = current.next if current == @head
        @tail = current.prev if current == @tail

        @size -= 1 # Restamos una estación si la encontramos y eliminamos
        break
      end
      current = current.next
    end
  end
end